### ISSUE TITLE
Idea: Escrowless Staking Mechanic via Transfer Hooks

### ISSUE THREAD

**Issue by amilz** [wen-community/wen-program-library#25]
## TLDR
Proposal to add staking to NFTs via a transfer hook following a similar pattern to royalties, structured as a `staked` boolean in the NFT's metadata.

Happy to take a stab at a PR but wanted to take a temperature check before running with something. 

## Problem
Many projects need a simple way to allow their token holders to stake their NFTs to verify that they are not actively listing/selling their NFTs. Because no standard exist, many projects have to build or buy their own staking solutions.

Additionally, most staking protocols require token transfers to escrow or delegating authorities. Though this is generally okay, it is challenging for doing queries by owners or for tracking NFTs across multiple protocols.

## Changes Required
A new metadata field, `staked` will need to be created and toggleable. Additionally the Transfer hook will need to be updated to check if the NFT is staked before allowing a transfer.

### Changed Files: 
- `src/instructions/royalty/execute.rs`
- Probably migrate `src/instructions/royalty/execute.rs` to `src/instructions/transfer_hook/execute_transfer_hook.rs`. 

#### Sample Implementation of `execute_transfer_hook.rs`

```rs
pub fn handler(ctx: Context<ExecuteTransferHook>) -> Result<()> {
    ctx.accounts.check_staking()?;
    ctx.accounts.check_royalties()?;

    Ok(())
}

impl<'info> ExecuteTransferHook<'info> {
    pub fn check_staking(&self) -> Result<()> {
        let metadata = get_metadata(&ctx.accounts.mint)?;

        // Check if the "Staked" field exists and is set to "true"
        if let Some(staked) = metadata.get("Staked") {
            if staked == "true" {
                return Err(error!(ExecuteTransferHookError::AssetStaked));
            }
        }
        Ok(())

        }

    pub fn check_royalties(&self) -> Result<()> {
        // Existing check on royalties
        Ok(())
    }
}
```

### New Insctructions/Files: 
- `src/instructions/mint/staking/add.rs` (very similar to `royalties/add.rs`)
- `src/instructions/mint/staking/mod.rs` 
- `src/instructions/staking/update/stake.rs`
- `src/instructions/staking/update/unstake.rs`
- `src/instructions/staking/update/mod.rs`

#### Concept for Add Staking Instruction

```rs
impl<'info> AddRoyalties<'info> {
    fn update_token_metadata_field(&self, field: Field, value: String) -> ProgramResult {
        // Same as adding royalties
        Ok(())
    }

    fn update_transfer_hook_program_id(&self) -> Result<()> {
        // Same as adding royalties
        Ok(())
    }
}

pub fn handler(ctx: Context<AddRoyalties>, args: AddRoyaltiesArgs) -> Result<()> {
    // add royalty basis points to metadata
    ctx.accounts.update_token_metadata_field(
        Field::Key(STAKING_FIELD.to_owned()),
        "false".to_string(),
    )?;

    // Depending on setup, may or may not requre adding a PDA to extra metas list
    
    // add metadata program as the transfer hook program
    ctx.accounts.update_transfer_hook_program_id()?;

    // Update minimimum balances (Same as adding royalties)

    Ok(())
}

```

To stake and unstake, we can utilize `update_token_metadata_field`. 

#### New Errors: 

```rs
#[error_code]
pub enum ExecuteTransferHookError {
    #[msg("The asset is currently staked and cannot be transferred.")]
    AssetStaked,
    #[msg("The asset is currently staked and cannot be staked again.")]
    AssetAlreadyStaked,
    #[msg("The asset is not currently staked and cannot be unstaked.")]
    AssetNotStaked,
}
```

## Considerations

- Nice for tracking NFTs (don't have to look into escrows or other programs to see if an NFT is staked); standardized
- No new account metas required.
- No new approvals required since metadata boolean is adequate to pass or fail the check.
- You're not really staking the NFT "to" anything, so multiple protocols could utilize this. IT'S a simple options for less-dev heavy NFT teams. You can easily add a stake/unstake button. Do staked snapshots. And since it's tied to transfer hook...you can be certain the NFT isn't being listed, sold, or transferred while staked.

### Do Nothing
This is okay--everybody can DIY their own thing. 

### PubKey over Boolean
Alternatively...instead of true...you could have a pubkey that the NFT is staked to. This could allow for a more complex staking system and prevent collisions between two programs rewarding a user for staking the same NFT.

If it's a pubkey, we'd need that key to sign to ensure that ppl aren't fake staking NFTs...eg i write a script to stake to foxy program and get $FOXY when unstaking.

Unlike most staking protocols you don't have a PDA storing time lock in this method. 

Could require a signer param e.g. on staking (and open for unstaking...though you might lose some protocol points).

```rs
#[account(mut, signer)]
staked_to: AccountInfo<'info>,
```

### PDA for Staking Terms

- Make a PDA inside WNS that holds common stake terms (seeded on MINT and "STAKE" or something else similar to the APPROVE PDA) 
- Struct could be like: 

```rs
pub struct StakingTerms {
    pub staked_to: Pubkey,
    pub staked_at: u64,
    pub last_claimed: u64,
    // etc...
}
```

**Comment by kespinola** [comment]
@amilz I see a lot of synergy with the token extension guard program we are currently working on. One issue with transfer hooks is that only 1 program can be registered at a time so a generalized guard framework is need to facilitate multiple use cases.

https://github.com/wen-community/wen-new-standard/issues/71

In this case the the guard would check the token and ensure a stake field is set and points to the desired value.

```rust
enum CPIRuleV1 {
  Allow(Vec<Pubkey>)
  Deny(Vec<Pubkey>)
}
// Control which protocols can interact with a mint's tokens. eg only allow royalty respecting protocols to facilitate transfers.
enum CPIRule {
  V1(CPIRulV1)
}

struct TransferAmountRuleV1 {
  Above(u64),
  Below(u64),
  Equal(u64),
  Rang(u64, u64)
}

/// Enforce the transfer amount is above, below, equal to, or within a range set my the mint authority.
enum TransferAmountRule {
   V1(TransferAmountRuleV1)
}


/// Ensure the token has 
struct Guard {
  cpi_rule: Option<CPIRule>,
  transfer_amount_rule: Option<TransferAmountRule>,
  token_state_rule: Option<TokenStateRule>
}
```

**Comment by kespinola** [comment]
The transfer guard program is now available in devnet. Its additional fields rules should meet this use case.
